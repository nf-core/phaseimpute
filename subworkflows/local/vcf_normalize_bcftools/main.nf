include { BCFTOOLS_NORM                         } from '../../../modules/nf-core/bcftools/norm'
include { BCFTOOLS_INDEX as BCFTOOLS_INDEX_1    } from '../../../modules/nf-core/bcftools/index'
include { BCFTOOLS_INDEX as BCFTOOLS_INDEX_2    } from '../../../modules/nf-core/bcftools/index'
include { BCFTOOLS_INDEX as BCFTOOLS_INDEX_3    } from '../../../modules/nf-core/bcftools/index'
include { BCFTOOLS_VIEW                         } from '../../../modules/nf-core/bcftools/view'
include { VCFLIB_VCFFIXUP                       } from '../../../modules/nf-core/vcflib/vcffixup/main'


workflow VCF_NORMALIZE_BCFTOOLS {
    take:
    ch_vcf          // channel: [ [id, chr], vcf, index ]
    ch_fasta        // channel: [ [genome], fasta, fai ]

    main:

    ch_versions = Channel.empty()
    ch_fasta = ch_fasta.map { meta, fasta, fai -> [meta, fasta] }

    // Join duplicated biallelic sites into multiallelic records
    BCFTOOLS_NORM(ch_vcf, ch_fasta)
    ch_versions = ch_versions.mix(BCFTOOLS_NORM.out.versions)

    // Index multiallelic VCF
    BCFTOOLS_INDEX_1(BCFTOOLS_NORM.out.vcf)
    ch_versions = ch_versions.mix(BCFTOOLS_INDEX_1.out.versions)

    // Join multiallelic VCF and TBI
    ch_multiallelic_vcf_tbi = BCFTOOLS_NORM.out.vcf.join(BCFTOOLS_INDEX_1.out.tbi)

    // Remove all multiallelic records and samples specified in the `--remove_samples` command:
    BCFTOOLS_VIEW(ch_multiallelic_vcf_tbi, [], [], [])
    ch_versions = ch_versions.mix(BCFTOOLS_VIEW.out.versions)

    // Index biallelic VCF
    BCFTOOLS_INDEX_2(BCFTOOLS_VIEW.out.vcf)
    ch_versions = ch_versions.mix(BCFTOOLS_INDEX_2.out.versions)

    // Join biallelic VCF and TBI
    ch_biallelic_vcf_tbi = BCFTOOLS_VIEW.out.vcf.join(BCFTOOLS_INDEX_2.out.tbi)

    // (Optional) Fix panel (When AC/AN INFO fields in VCF are inconsistent with GT field)
    if (params.compute_freq == true) {
        VCFLIB_VCFFIXUP(ch_biallelic_vcf_tbi)
        ch_versions = ch_versions.mix(VCFLIB_VCFFIXUP.out.versions)

        // Index fixed panel
        BCFTOOLS_INDEX_3(VCFLIB_VCFFIXUP.out.vcf)
        ch_versions = ch_versions.mix(BCFTOOLS_INDEX_3.out.versions)

        // Join fixed vcf and tbi
        ch_biallelic_vcf_tbi = VCFLIB_VCFFIXUP.out.vcf.join(BCFTOOLS_INDEX_3.out.tbi)
    }
    emit:
    vcf_tbi        = ch_biallelic_vcf_tbi            // channel: [ [id, chr], vcf, tbi ]
    versions       = ch_versions                    // channel: [ versions.yml ]
}
