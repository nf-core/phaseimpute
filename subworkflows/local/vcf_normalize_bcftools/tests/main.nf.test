nextflow_workflow {

    name "Test Subworkflow VCF_NORMALIZE_BCFTOOLS"
    script "../main.nf"

    config "./nextflow.config"

    workflow "VCF_NORMALIZE_BCFTOOLS"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/vcf_normalize_bcftools"
    tag "vcf_normalize_bcftools"

    tag "bcftools"
    tag "bcftools/norm"
    tag "bcftools/view"
    tag "bcftools/index"
    tag "vcflib"
    tag "vcflib/vcffixup"


    test("Normalize vcf without computing frequencies nor removing samples") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
                compute_freq = false
                remove_samples = null
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [id: "1000GP", chr: "chr22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz.csi", checkIfExist:true),
                    ],
                    [
                        [id: "1000GP", chr: "chr21"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21.s.norel.vcf.gz.csi", checkIfExist:true),
                    ]
                )
                input[1] = Channel.of([
                    [id: "GRCh38"],
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz", checkIfExist:true),
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz.fai", checkIfExist:true),
                ]).collect()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert snapshot(workflow.out.vcf_tbi.collect{
                    path(it[1]).vcf.summary
                    }).match("Normalize content")
                }
            )
        }
    }

    test("Normalize vcf with computing frequencies after removing samples") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
                compute_freq = true
                remove_samples = "HG00109,HG00110,HG00111,HG00112,HG00113,HG00114"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [id: "1000GP", chr: "chr22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz.csi", checkIfExist:true),
                    ],
                    [
                        [id: "1000GP", chr: "chr21"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21.s.norel.vcf.gz.csi", checkIfExist:true),
                    ]
                )
                input[1] = Channel.of([
                    [id: "GRCh38"],
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz", checkIfExist:true),
                    file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38.s.fa.gz.fai", checkIfExist:true),
                ]).collect()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert snapshot(workflow.out.vcf_tbi.collect{
                    path(it[1]).vcf.summary
                    }).match("Norm content recomputed")
                }
            )
        }
    }
}
