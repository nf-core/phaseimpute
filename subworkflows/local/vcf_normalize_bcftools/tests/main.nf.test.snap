{
    "Normalize content": {
        "content": [
            [
                "VcfFile [chromosomes=[chr21], sampleCount=3196, variantCount=836, phased=true, phasedAutodetect=true]",
                "VcfFile [chromosomes=[chr22], sampleCount=3196, variantCount=903, phased=true, phasedAutodetect=true]"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-18T15:21:05.213839279"
    },
    "Normalize vcf with computing frequencies after removing samples": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "1000GP_chr21_fixed.vcf.gz:md5,07239edd29a6c5a8ca1c3226d0f83de5",
                        "1000GP_chr21_fixed.vcf.gz.tbi:md5,388eefaf444c084970d27dff0478b212"
                    ],
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr22"
                        },
                        "1000GP_chr22_fixed.vcf.gz:md5,5d4579b182241f5aa812206ac81835d7",
                        "1000GP_chr22_fixed.vcf.gz.tbi:md5,9088246d10963b3cd8704715dc09c436"
                    ]
                ],
                "1": [
                    "versions.yml:md5,3bba75c8e94eb2271ad38e3a1ee06ad8",
                    "versions.yml:md5,3bba75c8e94eb2271ad38e3a1ee06ad8",
                    "versions.yml:md5,bdf2c1ea1cace63b301158243f3fd12f",
                    "versions.yml:md5,bdf2c1ea1cace63b301158243f3fd12f",
                    "versions.yml:md5,c1918a7ebe19e200e8981e12c8fd52d4",
                    "versions.yml:md5,c1918a7ebe19e200e8981e12c8fd52d4",
                    "versions.yml:md5,da9510d939fc9cb45b897127f0d5e490",
                    "versions.yml:md5,da9510d939fc9cb45b897127f0d5e490"
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "1000GP_chr21_fixed.vcf.gz:md5,07239edd29a6c5a8ca1c3226d0f83de5",
                        "1000GP_chr21_fixed.vcf.gz.tbi:md5,388eefaf444c084970d27dff0478b212"
                    ],
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr22"
                        },
                        "1000GP_chr22_fixed.vcf.gz:md5,5d4579b182241f5aa812206ac81835d7",
                        "1000GP_chr22_fixed.vcf.gz.tbi:md5,9088246d10963b3cd8704715dc09c436"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3bba75c8e94eb2271ad38e3a1ee06ad8",
                    "versions.yml:md5,3bba75c8e94eb2271ad38e3a1ee06ad8",
                    "versions.yml:md5,bdf2c1ea1cace63b301158243f3fd12f",
                    "versions.yml:md5,bdf2c1ea1cace63b301158243f3fd12f",
                    "versions.yml:md5,c1918a7ebe19e200e8981e12c8fd52d4",
                    "versions.yml:md5,c1918a7ebe19e200e8981e12c8fd52d4",
                    "versions.yml:md5,da9510d939fc9cb45b897127f0d5e490",
                    "versions.yml:md5,da9510d939fc9cb45b897127f0d5e490"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-22T15:51:30.725699084"
    },
    "Norm content recomputed": {
        "content": [
            [
                "VcfFile [chromosomes=[chr21], sampleCount=3190, variantCount=836, phased=true, phasedAutodetect=true]",
                "VcfFile [chromosomes=[chr22], sampleCount=3190, variantCount=903, phased=true, phasedAutodetect=true]"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-18T15:23:37.057826205"
    },
    "Normalize vcf without computing frequencies nor removing samples": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "1000GP_chr21_biallelic_snps.vcf.gz:md5,17746c80b9e02787f4cb950a3c2efa93",
                        "1000GP_chr21_biallelic_snps.vcf.gz.tbi:md5,006d0bec1050410aa64d427e2fad42e2"
                    ],
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr22"
                        },
                        "1000GP_chr22_biallelic_snps.vcf.gz:md5,817ebeceadcaea1e95f92c7b67659f28",
                        "1000GP_chr22_biallelic_snps.vcf.gz.tbi:md5,d3460d21cacfbb15b34289454e2922a8"
                    ]
                ],
                "1": [
                    "versions.yml:md5,c1918a7ebe19e200e8981e12c8fd52d4",
                    "versions.yml:md5,c1918a7ebe19e200e8981e12c8fd52d4",
                    "versions.yml:md5,da9510d939fc9cb45b897127f0d5e490",
                    "versions.yml:md5,da9510d939fc9cb45b897127f0d5e490"
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "1000GP_chr21_biallelic_snps.vcf.gz:md5,17746c80b9e02787f4cb950a3c2efa93",
                        "1000GP_chr21_biallelic_snps.vcf.gz.tbi:md5,006d0bec1050410aa64d427e2fad42e2"
                    ],
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr22"
                        },
                        "1000GP_chr22_biallelic_snps.vcf.gz:md5,817ebeceadcaea1e95f92c7b67659f28",
                        "1000GP_chr22_biallelic_snps.vcf.gz.tbi:md5,d3460d21cacfbb15b34289454e2922a8"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c1918a7ebe19e200e8981e12c8fd52d4",
                    "versions.yml:md5,c1918a7ebe19e200e8981e12c8fd52d4",
                    "versions.yml:md5,da9510d939fc9cb45b897127f0d5e490",
                    "versions.yml:md5,da9510d939fc9cb45b897127f0d5e490"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-22T15:51:02.997348191"
    }
}